idx/memory/write_queue.go

```diff
@@ -25,13 +27,14 @@ type WriteQueue struct {
 // in batches
 func NewWriteQueue(index *UnpartitionedMemoryIdx, maxDelay time.Duration, maxBuffered int) *WriteQueue {
 	wq := &WriteQueue{
-		archives:    make(map[schema.MKey]*idx.Archive),
-		shutdown:    make(chan struct{}),
-		done:        make(chan struct{}),
-		maxBuffered: maxBuffered,
-		maxDelay:    maxDelay,
-		flushed:     make(chan struct{}, 1),
-		idx:         index,
+		archives:     make(map[schema.MKey]*idx.Archive),
+		shutdown:     make(chan struct{}),
+		done:         make(chan struct{}),
+		maxBuffered:  maxBuffered,
+		maxDelay:     maxDelay,
+		flushTrigger: make(chan struct{}, 1),
+		flushPending: false,
```

Comment:
This is a large change in WriteQueue behavior to make `maxBuffered` more of a heuristic than an absolute max. This was done to take more control of the locking in the `flush` function.